src/java.base/share/classes/javax/crypto/spec/HKDFParameterSpec.java

```diff
@@ -0,0 +1,483 @@
+ * AlgorithmParameterSpec kdfParameterSpec =
+ *             HKDFParameterSpec.ofExtract()
+ *                              .addIKM(label)
+ *                              .addIKM(ikm)
+ *                              .addSalt(salt).extractOnly();
+ *
+ *
+ *}
+ * {@snippet lang = java:
+ *
+ * // this usage depicts the initialization of an HKDF-Expand AlgorithmParameterSpec
+ * AlgorithmParameterSpec kdfParameterSpec =
+ *             HKDFParameterSpec.expandOnly(prk, info, 32);
+ *
+ *}
+ * {@snippet lang = java:
+ *
+ * // this usage depicts the initialization of an HKDF-ExtractExpand AlgorithmParameterSpec
+ * AlgorithmParameterSpec kdfParameterSpec =
+ *             HKDFParameterSpec.ofExtract()
+ *                              .addIKM(ikm)
+ *                              .addSalt(salt).thenExpand(info, 32);
+ *
+ *}
+ *
+ * @since 24
+ */
+@PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+public interface HKDFParameterSpec extends AlgorithmParameterSpec {
+
+    /**
+     * This {@code Builder} builds {@code Extract} and {@code ExtractThenExpand}
+     * objects.
+     * <p>
+     * The {@code Builder} is initialized via the {@code ofExtract} method of
+     * {@code HKDFParameterSpec}. As stated in the class description,
+     * {@code addIKM} and/or {@code addSalt} may be called as needed. Finally,
+     * the object is "built" by calling either {@code extractOnly} or
+     * {@code thenExpand} for {@code Extract} and {@code ExtractThenExpand}
+     * use-cases respectively. Note that the {@code Builder} is not
+     * thread-safe.
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.KEY_DERIVATION)
+    final class Builder {
+
+        List<SecretKey> ikms = new ArrayList<>();
+        List<SecretKey> salts = new ArrayList<>();
+
+        private Builder() {}
+
+        /**
+         * Creates a {@code Builder} for an {@code Extract}.
+         *
+         * @return a {@code Builder} to mutate
+         */
+        private Builder createBuilder() {
+            return this;
+        }
+
+        /**
+         * Builds an {@code Extract} from the current state of the
+         * {@code Builder}.
+         *
+         * @return an immutable {@code Extract}
+         */
+        public Extract extractOnly() {
+            return new Extract(ikms, salts);
+        }
+
+        /**
+         * Builds an {@code ExtractThenExpand}.
+         *
+         * @param info
+         *     the optional context and application specific information (may be
+         *     {@code null}); the byte array is copied to prevent subsequent
+         *     modification
+         * @param length
+         *     the length of the output key material
+         *
+         * @return an {@code ExtractThenExpand}
+         *
+         * @throws IllegalArgumentException
+         *     if {@code length} is not greater than 0
+         */
+        public ExtractThenExpand thenExpand(byte[] info, int length) {
```

Comment:
If it is a fairly common use case to not have an `info` parameter, then I think we should add a method that does not take this parameter, i.e. `thenExpand(int length)`.

Reply:
Every time HKDF-Expand is used either TLS 1.3 or HPKE, there is always a non-empty info.

Reply:
Ok, then I think the examples should not use "null" as the info value.

Reply:
Ok, I'll accept that I guess. Whenever I see "null" as a valid parameter value, I try to think if adding an overloaded method w/o that value is a better way to support it.